### ISSUE TITLE
Static or rarely-updated layers: content discard notification?

### ISSUE THREAD

**Issue by Artyom17** [immersive-web/layers#106]
This issue is related to #101 and #49, and even to old PR https://github.com/immersive-web/webxr/commit/9d3ccaf24cd202704dcd7d9c00c97b03129abef6

The situation is the following. The textures returned by the layers (via `getSubImage` or `getViewSubImage`) can be textures, directly used by the underlying VR compositor (it is so for Oculus). This way we avoid any extra copies and it is the most efficient way to render.

However, such textures might be freed / reallocated by the compositor under-the-hood. An example of such situation is when the VR device falls asleep/wakes up or when browser temporarily switches context to OS (some kind of dialog, settings, etc) and then gets back. Note, in both cases we should receive **onvisibilitychange** event. But I am not sure if there are other situations possible, which won't trigger the `onvisibilitychange` event.
The textures got re-allocated once the device wakes up or context is switches back.

It is not a problem for the layers which are updated every frame. However, it is an issue for the 'static' layers (#49) or for the layers which are not updated every frame ('semi-static'). Such layers will lose their content and they should get re-populated. 

It is not a [good] option to let UA preserve the content of such textures, because not all textures are allowed to be readout, especially if we are talking about MSAA ones.

Thus, I think, user's experience should get notification about such 'discard' event. I see two options:
* Utilize the **onvisibilitychange** and make it mandatory for the experience to handle it and to re-populate all the static and semi-static layers, or
* introduce a new event, like **ontexturesdiscard** (in session) which will target this kind of situations w/o linking it to visibility status and ask user's code to re-populate textures in this handler. This is a more universal solution, imo.

Any thoughts?

**Comment by cabanier** [comment]
This looks a lot like the [context lost](https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.15.2) event from WebGL.

I think we should have a similar event on the session to signal the author that all layers must be repopulated.

**Comment by Artyom17** [comment]
Yeah. However, you actually can't (I think) to re-populate the layers IN the event handler since getSubImage & getViewSubImage work only inside the rAF (right?)

**Comment by Artyom17** [comment]
So, one more potential solution is just to have a layer's method 'bool wasDiscarded()' that should be checked at the beginning of each rAF. The 'discarded' status got reset after first call to getSubImage or getViewSubImage. We still can have the event, though.

**Comment by cabanier** [comment]
> Yeah. However, you actually can't (I think) to re-populate the layers IN the event handler since getSubImage & getViewSubImage work only inside the rAF (right?)

Correct. After the event fires, the author should plan on repopulating all the layers.

**Comment by cabanier** [comment]
> So, one more potential solution is just to have a layer's method 'bool wasDiscarded()' that should be checked at the beginning of each rAF. The 'discarded' status got reset after first call to getSubImage or getViewSubImage. We still can have the event, though.

That could be very useful! @toji, thoughts?

**Comment by cabanier** [comment]
I added a new even to the WebXR spec to signal when the swapchain is reallocated: https://github.com/immersive-web/webxr/pull/1009
Once that lands, we can update the layers spec (if needed).

**Comment by toji** [comment]
Sorry I'm a bit late to this conversation. Definitely agree that we need a way to signal that static layers need to be repopulated, but I'm not sure immersive-web/webxr#1009 suggests the right thing.

> one more potential solution is just to have a layer's method 'bool wasDiscarded()' that should be checked at the beginning of each rAF.

I'm always a bit skeptical of booleans flags that flip to true and then back again for one frame only, but I get why the event alone may not be desirable because it doesn't allow developers to react at the time they receive the signal. I wonder if attaching a discarded flag to `XRFrame` would be a slightly cleaner pattern. That way you're still able to check for it at the right time, but we don't have to be concerned about the timing of setting/clearing a persistent flag on the layer/binding/session.

**Comment by cabanier** [comment]
I believe an event is the correct way to go because this could happen in the middle of a requestAnimationFrame call.
@Artyom17, what would happen if the swapchain goes away mid-draw? Maybe the entire context should be marked as lost/restored...

**Comment by Artyom17** [comment]
Swapchain can't go away in mid-frame. At least for Oculus. If this happens for other vendors then yeah, the whole context probably must be marked as lost.

**Comment by Artyom17** [comment]
> I'm always a bit skeptical of booleans flags that flip to true and then back again for one frame only, but I get why the event alone may not be desirable because it doesn't allow developers to react at the time they receive the signal. I wonder if attaching a discarded flag to `XRFrame` would be a slightly cleaner pattern. That way you're still able to check for it at the right time, but we don't have to be concerned about the timing of setting/clearing a persistent flag on the layer/binding/session.

The idea behind having the boolean flags per layer is that this way we might know if all the layers got updated or not. These per-layer flags get reset only once getSubImage / getViewSubImage is called, thus UA can check if all layers got updated at the end of the frame and if not - print a warning into a console (at least).
More over, the same flag can be used to check if the layer was ever populated (after creation).
This way we don't have to have that "booleans flags that flip to true and then back again for one frame only" at all; instead, we have proper states of each layer's content ("populated" / "not populated").

**Comment by toji** [comment]
> I believe an event is the correct way to go because this could happen in the middle of a requestAnimationFrame call.

Events can't be processed in the middle of a rAF call. Even if we detected the context loss mid-rAF we wouldn't be able to notify the page about it until the rAF is finished due to the way events are queued.

> Maybe the entire context should be marked as lost/restored...

This is an interesting idea, as it would effectively force developers to re-submit their imagery, but it's an incredibly heavy hammer to wield considering the desired outcome. I'd view it as a last resort "nuke it from orbit, it's the only way to be sure" kind of mechanism, and I doubt we'll have to get that drastic.

> Maybe, we should name the flag as "hasValidContent" instead of "wasDiscarded".

Oh! I like this direction. Especially because we were talking about reprojection behavior across vendors previously and found that some AR platforms don't support it. In fact, after that call I found that Chrome's mobile AR implementation doesn't support reprojection either, and if you don't submit new imagery every frame the content just disappears.

So, given that, maybe the best framing is something along the lines of a layer flag called `canReproject`, although I don't think that's an entirely accurate term for things like quad layers so maybe Artem's `hasValidContent` is more general purpose? And I do think this would be a valuable signal even for the exisiting `XRWebGLLayer`, though in the context of the behavior listed above I think we do want it to be a per-layer signal, since you may have some platforms that can handle static quad layers but require projection layers to be refreshed every frame.

**Comment by Artyom17** [comment]
Quads and all other layers are indeed reprojected as well (their poses get updated accordingly to the latest head pose; well, at least for Oculus). But 'canReproject' still sounds too specific (this indicates consequence, not the cause, imo).

**Comment by cabanier** [comment]
> Quads and all other layers are indeed reprojected as well (their poses get updated accordingly to the latest head pose; well, at least for Oculus). But 'canReproject' still sounds too specific (this indicates consequence, not the cause, imo).

Maybe `needsRedraw`.
It would be true for Oculus if the swapchain goes away.
It would always be true for Chrome AR.
It would be true for compositors that only reproject for a short while and then switch away (ie Magic Leap)

**Comment by toji** [comment]
> Maybe 'needsRedraw'...

I like that quite a bit!

**Comment by cabanier** [comment]
> > Maybe 'needsRedraw'...
> 
> I like that quite a bit!

Should this be part of the layers spec, or should it go in WebXR?
It's obvious for layers but it affects XRWebGLLayer as well.

**Comment by toji** [comment]
It could go on `XRLayer` if we wanted, but I could also make an argument for saying basically that we should treat `XRWebGLLayer` as if it always `needsRedraw`.

**Comment by cabanier** [comment]
Does it always need a redraw? It's certainly reprojected on Oculus, Chromium and Magic Leap.

I thought the boolean would become true if it could no longer be reprojected( ie too much head movement or the swapchain was lost.)

**Comment by toji** [comment]
It wouldn't strictly always need redraw, no, but considering the behavior that we see in existing implementations you kind of have to treat it as such.

**Comment by cabanier** [comment]
ok. Then it makes more sense to put it on the `XRCompositionLayer`.

**Comment by RafaelCintron** [comment]
If the purpose of the `needsRedraw` attribute is to inform developers they need to refresh infrequently accessed resources, it should be part of a new or existing notification. There may not be a continuous raF loop running to check the attribute.  

If we haven't already done so, we should spec that any WebGL textures returned by the WebXR APIs are only valid during the WebXR raF callback. Outside of the raF, the texture should act as if someone called `destroy` on texture. 

Doing this will avoid portability issues with developers holding a reference to the WebXR texture and using it outside of the raF, counting on contents containing certain bits.

**Comment by cabanier** [comment]
> If the purpose of the `needsRedraw` attribute is to inform developers they need to refresh infrequently accessed resources, it should be part of a new or existing notification. There may not be a continuous raF loop running to check the attribute.

Yes, this is is why I originally proposed an event.
However, now that we defined the attribute to signal both a reallocation of the swapchain and request for redraw from the compositor, it makes less sense because for some UAs (ie chrome AR) there would be an event for each raf call.

I think having both an event and an attribute might be confusing for authors...

> If we haven't already done so, we should spec that any WebGL textures returned by the WebXR APIs are only valid during the WebXR raF callback. Outside of the raF, the texture should act as if someone called `destroy` on texture.

That's already in the spec. See the last paragraph of this [section](https://immersive-web.github.io/layers/#xrwebglsubimagetype). 
Do you have a suggestion to make it more clear?

**Comment by toji** [comment]
> If the purpose of the `needsRedraw` attribute is to inform developers they need to refresh infrequently accessed resources, it should be part of a new or existing notification. There may not be a continuous raF loop running to check the attribute.

That's a good point. We expect that anyone with a projection layer (which should be the large majority of applications) will have a continuous rAF loop, but I can imagine a very simple image gallery app that doesn't bother drawing any UX and advances images on select events which could operate entirely with non-projection layers.

This would be a terrible app to use, and maybe that's a sufficient excuse for us to say "not supported.", but it's worth contemplating at the very least.

**Comment by RafaelCintron** [comment]
> I think having both an event and an attribute might be confusing for authors...

With just an attribute, how is the "static" or "semi-static" content going to find out that `needsRedraw` has been set to true and needs to be handled? 
 
> That's already in the spec. See the last paragraph of this section.
Do you have a suggestion to make it more clear?

Ah. I was looking at the main XR spec and didn't see it mentioned there.  I should have looked in the layers spec!  :-)

The section you refer to says that textures becomes "invalid" outside of the WebXR raF.  What does that mean in practice?  Today, it is possible for a web developer to delete a WebGL texture and still use it in framebuffers that are not currently bound but have the texture attached.

**Comment by cabanier** [comment]
> This would be a terrible app to use, and maybe that's a sufficient excuse for us to say "not supported.", but it's worth contemplating at the very least.

It wouldn't necessarily be a terrible app. For instance, one could create a video player that has a skybox implemented as a cubelayer. Controllers would be hidden by default unless a button is pressed. Such an app would have no need to draw anything in the rAF loop.

**Comment by cabanier** [comment]
> > I think having both an event and an attribute might be confusing for authors...
> 
> With just an attribute, how is the "static" or "semi-static" content going to find out that `needsRedraw` has been set to true and needs to be handled?

The author would have to check each layer during each rAF call

> > That's already in the spec. See the last paragraph of this section.
> > Do you have a suggestion to make it more clear?
> ...
> The section you refer to says that textures becomes "invalid" outside of the WebXR raF. What does that mean in practice? Today, it is possible for a web developer to delete a WebGL texture and still use it in framebuffers that are not currently bound but have the texture attached.

I don't know :-)
We probably need more specific text like we have in the [WebXR spec](https://immersive-web.github.io/webxr/#opaque-framebuffer).
For instance, do we have to reattach textures to framebuffers for each rAF call or does it all magically work?

**Comment by cabanier** [comment]
@RafaelCintron I think we should create another issue to cover the lifetime of the WebGL texture. What do you think?

**Comment by RafaelCintron** [comment]
@cabanier , I agree we should create another issue to cover WebGL texture lifetime issues.  

However, I think we should reconsider whether we should have an additional event (or piggy back onto an existing event) for these re-allocated textures.  You yourself said (and I agree) non-terrible apps will exist with no rAF loops.

**Comment by cabanier** [comment]
> @cabanier , I agree we should create another issue to cover WebGL texture lifetime issues.

I filed #111. I will talk to @Artyom17 to get a better understanding how things work on our side.

> However, I think we should reconsider whether we should have an additional event (or piggy back onto an existing event) for these re-allocated textures. You yourself said (and I agree) non-terrible apps will exist with no rAF loops.

What do you think if https://github.com/immersive-web/webxr/pull/1009?
